File: go-backend-interview-prep/07_io_files_networking/08_http_client_basics/main.py
"""HTTP client basics -- Python equivalent of the Go example.

Uses http.server for a local test server and urllib.request for client.
"""

import json
import threading
import time
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/hello":
            self.send_response(200)
            self.send_header("X-Custom", "demo-value")
            self.end_headers()
            self.wfile.write(f"Hello from local server! Method=GET".encode())
        elif self.path == "/json":
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"status": "ok", "count": 42}).encode())
        elif self.path == "/slow":
            time.sleep(3)
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b"slow response")
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, format, *args):
        pass  # suppress logs


def main() -> None:
    # --- Spin up local server ---
    server = HTTPServer(("127.0.0.1", 0), Handler)
    port = server.server_address[1]
    base_url = f"http://127.0.0.1:{port}"

    t = threading.Thread(target=server.serve_forever, daemon=True)
    t.start()
    print(f"local server at {base_url}")

    try:
        # --- Example 1: GET with timeout ---
        print("\n=== GET /hello ===")
        resp = urllib.request.urlopen(f"{base_url}/hello", timeout=5)
        body = resp.read().decode()
        print(f"  status: {resp.status}")
        print(f"  body: {body}")
        print(f"  X-Custom: {resp.headers.get('X-Custom')}")

        # --- Example 2: GET JSON ---
        print("\n=== GET /json ===")
        resp = urllib.request.urlopen(f"{base_url}/json", timeout=5)
        body = resp.read().decode()
        print(f"  Content-Type: {resp.headers.get('Content-Type')}")
        print(f"  body: {body}")

        # --- Example 3: custom request with headers ---
        print("\n=== Custom request with headers ===")
        req = urllib.request.Request(
            f"{base_url}/hello",
            headers={
                "Accept": "text/plain",
                "User-Agent": "py-interview-prep/1.0",
            },
        )
        resp = urllib.request.urlopen(req, timeout=5)
        print(f"  body: {resp.read().decode()}")

        # --- Example 4: timeout on slow endpoint ---
        print("\n=== Timeout (1s on /slow) ===")
        try:
            urllib.request.urlopen(f"{base_url}/slow", timeout=1)
        except Exception as e:
            print(f"  expected timeout: {e}")

    finally:
        server.shutdown()


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/07_io_files_networking/08_http_client_basics/main.go
package main

import (
	"fmt"
	"io"
	"net"
	"net/http"
	"time"
)

func main() {
	// --- Spin up a tiny local HTTP server ---
	mux := http.NewServeMux()
	mux.HandleFunc("/hello", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("X-Custom", "demo-value")
		fmt.Fprintf(w, "Hello from local server! Method=%s", r.Method)
	})
	mux.HandleFunc("/json", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"status":"ok","count":42}`))
	})
	mux.HandleFunc("/slow", func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(3 * time.Second) // intentionally slow
		w.Write([]byte("slow response"))
	})

	// Listen on a random port
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		fmt.Println("listen error:", err)
		return
	}
	baseURL := fmt.Sprintf("http://%s", listener.Addr().String())
	go http.Serve(listener, mux)
	defer listener.Close()

	fmt.Println("local server at", baseURL)

	// --- Example 1: GET with timeout ---
	fmt.Println("\n=== GET /hello ===")
	client := &http.Client{Timeout: 5 * time.Second}

	resp, err := client.Get(baseURL + "/hello")
	if err != nil {
		fmt.Println("  error:", err)
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Println("  status:", resp.StatusCode)
	fmt.Println("  body:", string(body))
	fmt.Println("  X-Custom:", resp.Header.Get("X-Custom"))

	// --- Example 2: GET JSON ---
	fmt.Println("\n=== GET /json ===")
	resp2, _ := client.Get(baseURL + "/json")
	defer resp2.Body.Close()
	body2, _ := io.ReadAll(resp2.Body)
	fmt.Println("  Content-Type:", resp2.Header.Get("Content-Type"))
	fmt.Println("  body:", string(body2))

	// --- Example 3: custom request with headers ---
	fmt.Println("\n=== Custom request with headers ===")
	req, _ := http.NewRequest("GET", baseURL+"/hello", nil)
	req.Header.Set("Accept", "text/plain")
	req.Header.Set("User-Agent", "go-interview-prep/1.0")

	resp3, _ := client.Do(req)
	defer resp3.Body.Close()
	body3, _ := io.ReadAll(resp3.Body)
	fmt.Println("  body:", string(body3))

	// --- Example 4: timeout on slow endpoint ---
	fmt.Println("\n=== Timeout (1s on /slow) ===")
	fastClient := &http.Client{Timeout: 1 * time.Second}
	_, err = fastClient.Get(baseURL + "/slow")
	if err != nil {
		fmt.Println("  expected timeout:", err)
	}
}
